Mathlib/Combinatorics/SimpleGraph/Copy.lean

```diff
@@ -356,6 +360,205 @@ alias ⟨IsIndContained.exists_iso_subgraph, IsIndContained.of_exists_iso_subgra
+`G`. By construction, it doesn't contain `H` (unless `H` had no edges) and has at most the number of
+copies of `H` edges less than `G`. -/
+noncomputable irreducible_def killCopies (G : SimpleGraph V) (H : SimpleGraph W) :
+    SimpleGraph V := by
+  classical exact
+  if hH : H = ⊥ then G
+  else G.deleteEdges <| ⋃ (G' : G.Subgraph) (hG' : Nonempty (H ≃g G'.coe)), {(aux hH hG').some}
+
+/-- Removing an edge from `G` for each subgraph isomorphic to `H` results in a subgraph of `G`. -/
+lemma killCopies_le_left : G.killCopies H ≤ G := by
+  rw [killCopies]; split_ifs; exacts [le_rfl, deleteEdges_le _]
+
+@[simp] lemma killCopies_bot (G : SimpleGraph V) : G.killCopies (⊥ : SimpleGraph W) = G := by
+  rw [killCopies]; exact dif_pos rfl
+
+private lemma killCopies_of_ne_bot (hH : H ≠ ⊥) (G : SimpleGraph V) :
+    G.killCopies H =
+      G.deleteEdges (⋃ (G' : G.Subgraph) (hG' : Nonempty (H ≃g G'.coe)), {(aux hH hG').some}) := by
+  rw [killCopies]; exact dif_neg hH
+
+lemma killCopies_eq_left (hH : H ≠ ⊥) : G.killCopies H = G ↔ H.Free G := by
+  simp only [killCopies_of_ne_bot hH, Set.disjoint_left, isContained_iff_exists_iso_subgraph,
+    @forall_swap _ G.Subgraph, Set.iUnion_singleton_eq_range, deleteEdges_eq_self, Set.mem_iUnion,
+    Set.mem_range, not_exists, not_nonempty_iff, Nonempty.forall, Free]
+  exact forall_congr' fun G' ↦ ⟨fun h ↦ ⟨fun f ↦ h _
+    (Subgraph.edgeSet_subset _ <| (aux hH ⟨f⟩).choose_spec) f rfl⟩, fun h _ _ ↦ h.elim⟩
+
+protected lemma Free.killCopies_eq_left (hHG : H.Free G) : G.killCopies H = G := by
+  obtain rfl | hH := eq_or_ne H ⊥
+  · exact killCopies_bot _
+  · exact (killCopies_eq_left hH).2 hHG
+
+/-- Removing an edge from `G` for each subgraph isomorphic to `H` results in a graph that doesn't
+contain `H`. -/
+lemma free_killCopies (hH : H ≠ ⊥) : H.Free (G.killCopies H) := by
+  rw [killCopies_of_ne_bot hH, deleteEdges, Free, isContained_iff_exists_iso_subgraph]
+  rintro ⟨G', hHG'⟩
+  have hG' : (G'.map <| .ofLE (sdiff_le : G \ _ ≤ G)).edgeSet.Nonempty := by
+    rw [Subgraph.edgeSet_map]
+    exact (aux hH hHG').image _
+  set e := hG'.some with he
+  have : e ∈ _ := hG'.some_mem
+  clear_value e
+  rw [← Subgraph.image_coe_edgeSet_coe] at this
+  subst he
+  obtain ⟨e, he₀, he₁⟩ := this
+  let e' : Sym2 G'.verts := Sym2.map (Copy.isoSubgraphMap (.ofLE _ _ _) _).symm e
+  have he' : e' ∈ G'.coe.edgeSet := (Iso.map_mem_edgeSet_iff _).2 he₀
+  rw [Subgraph.edgeSet_coe] at he'
+  have := Subgraph.edgeSet_subset _ he'
+  simp only [edgeSet_sdiff,  edgeSet_fromEdgeSet,  edgeSet_sdiff_sdiff_isDiag, Set.mem_diff,
+    Set.mem_iUnion, not_exists] at this
+  refine this.2 (G'.map <| .ofLE sdiff_le) ⟨((Copy.ofLE _ _ _).isoSubgraphMap _).comp hHG'.some⟩ ?_
+  rw [Sym2.map_map, Set.mem_singleton_iff, ← he₁]
+  congr 1 with x
+  exact congr_arg _ (Equiv.Set.image_symm_apply _ _ injective_id _ _)
+
+variable [Fintype G.edgeSet]
+
+noncomputable instance killCopies.EdgeSet.fintype : Fintype (G.killCopies H).edgeSet :=
```

Comment:
I think the naming convention says instFintype for this

Reply:
Ah yes indeed